src/irep2/irep2_utils.h

```diff
@@ -382,4 +401,219 @@ inline expr2tc gen_one(const type2tc &type)
   abort();
 }
 
+/**
+   * @brief Distribute the functor `func` over op1 and op2
+   * at least one of those must be a vector
+   *
+   * @param func the functor operation e.g add, sub, mul
+   * @param op1 the first operand
+   * @param op2 the second operand
+   * @return expr2tc with the resulting vector
+   */
+inline expr2tc distribute_vector_operation(
+  std::function<expr2tc(type2tc, expr2tc, expr2tc)> func,
+  expr2tc op1,
+  expr2tc op2)
```

Comment:
1. Unlike the equally-named function below, the `op2` parameter here is not defaulted to the nil expression. My understanding of it is that a nil op2 is supported as well. Could it be made clear in the comment(s), whether and if so when that's the case?
2. As this function is inline and exposed in a header already, could we make it a template taking `<typename Func>` instead of `std::function` in order to avoid dynamic allocations and dynamic dispatch for lambdas? This would be useful for e.g. the vector-related code in simplify_expr2.cpp.

Reply:
I've implemented 2.

For item 1... the only restriction is the `func` should support whichever is given as op1/op2. I am not sure whether we can encode this as an assertion though.

Reply:
No need for an assertion, I just wanted the comment on this function to be updated to say what's going with op1 and op2.

Reply:
Actually 2. is not implemented, yet :) now there's an implicit conversion to `std::funtion` at every call site. What I meant was for the signature of this function to be
```c++
template <typename Func>
inline expr2tc distribute_vector_operation(
  Func func,
  expr2tc op1,
  expr2tc op2)
```

Reply:
I've changed the function signature and added the comment about nil in the documentation

Reply:
Great, thank you!